### DISCUSSION TITLE
readContracts freeze my react app

### DISCUSSION THREAD

**Discussion by bahoss** [wevm/wagmi#2729] [Q&A]
Hi,
I need to fetch balances by 5000> tokens,  but my app freeze. I use action readContracts
```js
const tokensBalances = await readContracts({
      contracts: tokens!.map(({ contractAddress }) => ({
        addressOrName:  contractAddress,
        functionName: 'balanceOf',
        contractInterface: ABI.ERC20,
        args: [address],
      })),
    });
```

Can anyone help? what i am doing wrong(

**Comment by gradaom** [comment]
Here's an updated version of your code with some optimizations and error handling:

```javascript
try {
  // Create an array to store the promises for fetching balances
  const balancePromises = tokens.map(({ contractAddress }) =>
    readContracts({
      contracts: [
        {
          addressOrName: contractAddress,
          functionName: 'balanceOf',
          contractInterface: ABI.ERC20,
          args: [address],
        },
      ],
    })
  );

  // Execute the promises concurrently using Promise.all
  const tokenBalances = await Promise.all(balancePromises);

  // Extract the balances from the results
  const balances = tokenBalances.map(result => result[0].toNumber());

  console.log('Balances:', balances);
} catch (error) {
  console.error('Error fetching balances:', error);
}
```

In this updated code:

1. We use `Promise.all` to execute the `readContracts` promises concurrently, which can improve performance by fetching balances in parallel.

2. The results of the promises are stored in the `tokenBalances` array.

3. We extract the balances from the results using `map` and convert them to numbers if needed.

4. If any error occurs during the execution, it will be caught in the `catch` block and logged to the console.

Please note that this code assumes you have properly defined `tokens`, `readContracts`, `address`, and `ABI.ERC20`. Make sure these variables are correctly initialized before using this code.

I hope this helps! Let me know if you have further questions.
